test/common/runtime/runtime_impl_test.cc

```diff
@@ -1036,6 +1045,39 @@ TEST_F(RtdsLoaderImplTest, DeltaOnConfigUpdateSuccess) {
   EXPECT_EQ(2, store_.gauge("runtime.num_layers", Stats::Gauge::ImportMode::NeverImport).value());
 }
 
+// Delta style add and removal successful update.
+TEST_F(RtdsLoaderImplTest, DeltaOnConfigUpdateWithRemovalSuccess) {
+  setup();
+
+  auto runtime = TestUtility::parseYaml<envoy::service::runtime::v3::Runtime>(R"EOF(
+    name: some_resource
+    layer:
+      foo: bar
+      baz: meh
+  )EOF");
+  EXPECT_CALL(rtds_init_callback_, Call());
+  doDeltaOnConfigUpdateVerifyNoThrow(runtime);
+
+  EXPECT_EQ("bar", loader_->snapshot().get("foo").value().get());
+  EXPECT_EQ("yar", loader_->snapshot().get("bar").value().get());
+  EXPECT_EQ("meh", loader_->snapshot().get("baz").value().get());
+
+  EXPECT_EQ(0, store_.counter("runtime.load_error").value());
+  EXPECT_EQ(2, store_.counter("runtime.load_success").value());
+  EXPECT_EQ(3, store_.gauge("runtime.num_keys", Stats::Gauge::ImportMode::NeverImport).value());
+  EXPECT_EQ(2, store_.gauge("runtime.num_layers", Stats::Gauge::ImportMode::NeverImport).value());
+
+  doDeltaOnConfigRemovalVerifyNoThrow("some_resource");
+
+  EXPECT_EQ("whatevs", loader_->snapshot().get("foo").value().get());
```

Comment:
Would be nice to verify that an attempt to read the keys from the layer we cleared out (bar/meh) no longer returns anything

Reply:
the existing keys at this point are:
`foo` `bar` - both from the bottom layer and came from `setup()` on L1050, value for these two keys at this point are 
`foo`->`whatevs`
`bar` -> `yar`

`foo` had a value before L1070 and is verified on L1061. `bar` had the same value, because it's not overwritten in the upper layer.

key `baz` does not exist anymore, and an attempt to access this key will yield an error: 
<img width="1500" alt="Screen Shot 2020-10-16 at 5 59 44 PM" src="https://user-images.githubusercontent.com/57155915/96625626-4aee8380-12c3-11eb-8ead-e9a82acfb7cc.png">

the removal of `baz` is verified by the number of keys reported from the gauge on L1077.

Reply:
The bad optional access is coming from valling `value()` on the return value, you should be able to verify that its returning empty by checking `has_value()`: https://en.cppreference.com/w/cpp/utility/optional/operator_bool